kernel/src/syscall.rs

```diff
@@ -222,8 +222,8 @@ pub enum SyscallReturnVariant {
 /// method.
 ///
 /// Capsules do not use this struct. Capsules use higher level Rust types
-/// (e.g. [`ReadWriteAppSlice`](crate::ReadWriteAppSlice) and
-/// `UpcallMemory`) or wrappers around this struct
+/// (e.g. [`ReadWriteProcessBuffer`](crate::ReadWriteProcessBuffer) and
+/// [`UpcallMemory`]) or wrappers around this struct
```

Comment:
Unfortunately I don't think this will work as UpcallData is not public.

Reply:
Right, I just realized.

Reply:
Can't we just reference the `GrantUpcallTable` type here? After all, that's what capsules are confronted with?

Reply:
Yeah that's the type I meant. I tried before and it didn't work, but I could have missed something or it could have changed.

Reply:
Oh, sorry. I was confused, as we can't have a private struct in a public interface, but there's nothing stopping us from just not exporting that.

I suppose this leads to a different question, which is perhaps best answered in #2551: should we export these types? It makes cross-referencing them in documentation impossible, and more importantly does not allow the user to introspect their public API. For example, take a look at the `Grant::enter` method [as it is currently documented](https://docs.tockos.org/kernel/struct.grant#method.enter). It's a bummer I can't click on the `GrantUpcallTable` and see that I have to call it's `schedule_upcall` method, but instead have to look into the source.

Reply:
Ack Rust tools are always fighting against safety! The problem is making things public can lead to soundness issues (which used to be a much bigger issue than it is today, and is what lead to Capabilities) and it's often hard to reason about 1) why a thing in kernel/ is public at all and 2) if it is public is it safe (or not).

I also don't understand why this would be considered a tock bug and not a cargo doc bug. If code in a different crate can use the api just fine, then why shouldn't cargo doc document it.

Reply:
Agreed. Will ask Rust folks or open an issue.